### PR TITLE
drivers: flash: Support for IFX QSPI Flash driver

### PR DESCRIPTION
**Pull request by sreeramIfx**
Initial version

### CODE REVIEW

File: drivers/flash/flash_ifx_cat1_qspi.c [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+static cy_stc_smif_hybrid_region_info_t *sfdp_slave_slot_0_region_info[16] = {
+	&sfdp_slave_slot_0_region_info_storage[0],  &sfdp_slave_slot_0_region_info_storage[1],
+	&sfdp_slave_slot_0_region_info_storage[2],  &sfdp_slave_slot_0_region_info_storage[3],
+	&sfdp_slave_slot_0_region_info_storage[4],  &sfdp_slave_slot_0_region_info_storage[5],
+	&sfdp_slave_slot_0_region_info_storage[6],  &sfdp_slave_slot_0_region_info_storage[7],
+	&sfdp_slave_slot_0_region_info_storage[8],  &sfdp_slave_slot_0_region_info_storage[9],
+	&sfdp_slave_slot_0_region_info_storage[10], &sfdp_slave_slot_0_region_info_storage[11],
+	&sfdp_slave_slot_0_region_info_storage[12], &sfdp_slave_slot_0_region_info_storage[13],
+	&sfdp_slave_slot_0_region_info_storage[14], &sfdp_slave_slot_0_region_info_storage[15],
```
Comment (LingaoM) [inline drivers/flash/flash_ifx_cat1_qspi.c:101]: use `LISTIFY`.

Comment (sreeramIfx) [reply drivers/flash/flash_ifx_cat1_qspi.c]: updated to use LISTIFY

File: drivers/flash/flash_ifx_cat1_qspi.c [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+};
+
+cy_stc_smif_octal_ddr_en_seq_t oe_sequence_SFDP_SlaveSlot_0 = {
+	.cmdSeq1Len = CY_SMIF_SFDP_ODDR_CMD_SEQ_MAX_LEN,
+	.cmdSeq2Len = CY_SMIF_SFDP_ODDR_CMD_SEQ_MAX_LEN,
+	.cmdSeq1 = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
+	.cmdSeq2 = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
+};
+
+#if (CY_SMIF_DRV_VERSION_MAJOR > 1) || (CY_SMIF_DRV_VERSION_MINOR >= 50)
```
Comment (ifyall) [inline drivers/flash/flash_ifx_cat1_qspi.c:90]: Shouldn't this logic be &&? Right now, we are selecting if the major version is greater than 1 OR the minor version is greater/equal to 50. So this selects driver 1.50 or greater. Is that what we want, or do we want v2.50 or greater? Either way, we should have a comment here explaining why we are choosing whatever minimum version we intend.

File: drivers/flash/flash_ifx_cat1_qspi.c [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+	.stsRegBusyMask = 0x00U,
+	/* The mask for the status register. */
+	.stsRegQuadEnableMask = 0x00U,
+	/* The max time for the erase type-1 cycle-time in ms. */
+	.eraseTime = 1U,
+	/* The max time for the chip-erase cycle-time in ms. */
+	.chipEraseTime = 16U,
+	/* The max time for the page-program cycle-time in us. */
+	.programTime = 8U,
+#if (CY_SMIF_DRV_VERSION_MAJOR > 1) || (CY_SMIF_DRV_VERSION_MINOR >= 50)
```
Comment (ifyall) [inline drivers/flash/flash_ifx_cat1_qspi.c:138]: Same version check as above, so if that logic had to change, change this logic as well..

File: modules/hal_infineon/mtb-hal-cat1/CMakeLists.txt [hunk 4]
```diff
@@ -86,7 +86,7 @@ zephyr_library_sources_ifdef(CONFIG_USE_INFINEON_TRNG    ${hal_dir}/source/cyhal
 zephyr_library_sources_ifdef(CONFIG_USE_INFINEON_UART    ${hal_dir}/source/cyhal_uart.c)
 zephyr_library_sources_ifdef(CONFIG_USE_INFINEON_WDT     ${hal_dir}/source/cyhal_wdt.c)
 
-if(CONFIG_USE_INFINEON_ADC)
+if(CONFIG_USE_INFINEON_ADC OR CONFIG_USE_INFINEON_SMIF)
```
Comment (ifyall) [inline modules/hal_infineon/mtb-hal-cat1/CMakeLists.txt:89]: This if statement is strange. Why do we need cyhal_analog_common.c when we use the SMIF?

**Review by ifyall — CHANGES_REQUESTED** [review]

File: boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts [hunk 5]
```diff
@@ -72,34 +72,45 @@ uart2: &scb2 {
...
+			bootstrap_region: bootstrap_region@60000050 {
 				compatible = "zephyr,memory-region", "soc-nv-flash";
 				zephyr,memory-region = "BOOTSTRAP_FLASH";
 				reg = <0x60000050 0x2550>;
 			};
-			app_region: flash@60002600 {
+
+			app_region: app_region@60002600 {
 				compatible = "soc-nv-flash";
 				reg = <0x60002600 0x5da00>;
```
Comment (npal-cy) [inline boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts:102]: can you use DT_SIZE_K(374)>; here as well

File: boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02_defconfig [hunk 6]
```diff
@@ -29,3 +29,6 @@ CONFIG_CLOCK_CONTROL=y
 
 # Main Stack Size
 CONFIG_MAIN_STACK_SIZE=2048
+
+# Enable core relocation ti move SMIF driver into RAM
```
[the remark this answers is not in the document]
Comment (sreeramIfx) [reply boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02_defconfig]: updated the comment

File: drivers/flash/flash_ifx_cat1_qspi.c (hunk 2 above)
Comment (sreeramIfx) [reply drivers/flash/flash_ifx_cat1_qspi.c]: It should have been `&&`, updated the logic. Support for memories with hybrid regions is added in the version 1.50, hence the condition. Also added a comment pointing to the changelog with more details.

File: drivers/flash/flash_ifx_cat1_qspi.c (hunk 3 above)
Comment (sreeramIfx) [reply drivers/flash/flash_ifx_cat1_qspi.c]: updated the logic.

File: modules/hal_infineon/mtb-hal-cat1/CMakeLists.txt (hunk 4 above)
Comment (sreeramIfx) [reply modules/hal_infineon/mtb-hal-cat1/CMakeLists.txt:90]: cyhal_analog_common.c is not required for SMIF, only DMA is required. Moved cyhal_analog_common.c out of this condition into a separate line

File: boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts (hunk 5 above)
Comment (sreeramIfx) [reply boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts:102]: already updated @ https://github.com/zephyrproject-rtos/zephyr/pull/74174

File: modules/hal_infineon/CMakeLists.txt [hunk 7]
```diff
@@ -19,6 +19,8 @@ if (CONFIG_SOC_FAMILY_INFINEON_CAT1 OR CONFIG_SOC_FAMILY_PSOC6_LEGACY)
   ## Add mtb-pdl-cat1 sources for CAT1 devices
   add_subdirectory(mtb-pdl-cat1)
 
+  add_subdirectory(serial-flash)
```
[the remark this answers is not in the document]
Comment (sreeramIfx) [reply modules/hal_infineon/CMakeLists.txt]: moved this into `if (CONFIG_SOC_FAMILY_INFINEON_CAT1 AND NOT CONFIG_SOC_FAMILY_PSOC6_LEGACY)
`

**Review by ifyall — DISMISSED** [review]

**Review by npal-cy — DISMISSED** [review]

**Review by ifyall — DISMISSED** [review]

File: drivers/flash/Kconfig [hunk 8]
```diff
@@ -220,6 +220,7 @@ source "drivers/flash/Kconfig.gd32"
 source "drivers/flash/Kconfig.xmc4xxx"
 
 source "drivers/flash/Kconfig.ifx_cat1"
+
```
Comment (npal-cy) [inline drivers/flash/Kconfig:223]: do we need this changes? if no please revert...

**Review by ifyall — APPROVED** [review]

**Review by npal-cy — APPROVED** [review]

**Pull request merged**